File: 3597. Partition String.cpp
/*
    Task:
        Process a string to get unique substrings while iterating it.

    Idea:
        Simulate the given algorithm using a set and an answer array
        to keep the order.

        Time complexity: O(n)
        Space complexity: O(n)
*/
class Solution {
public:
    vector<string> partitionString(string s) {
        unordered_set<string> auxSet;
        vector<string> uniqueSegments;

        string auxString = "";
        for(char c : s)
        {
            auxString += c;
            if(auxSet.find(auxString) == auxSet.end())
            {
                auxSet.insert(auxString);
                uniqueSegments.push_back(auxString);
                auxString = "";
            }
        }

        return uniqueSegments;
    }
};


/*
s = "abbccccd"
            "d"
     ["a", "b", "bc", "c", "cc", "d"]
*/

File: 3597. Partition String.java
/*
    Task:
        Process a string to get unique substrings while iterating it.

    Idea:
        Simulate the given algorithm using a set and an answer array
        to keep the order.

        Time complexity: O(n)
        Space complexity: O(n)
*/
class Solution {
    public List<String> partitionString(String s) {
        Set<String> auxSet = new HashSet<>();
        List<String> uniqueSegments = new ArrayList<>();

        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < s.length(); i++)
        {
            sb.append(s.charAt(i));
            String aux = sb.toString();
            if(!auxSet.contains(aux))
            {
                auxSet.add(aux);
                uniqueSegments.add(aux);
                sb.setLength(0);
            }
        }

        return uniqueSegments;
    }
}

/*
    s = "abbccccd"
               "d"
     ["a", "b", "bc", "c", "cc", "d"]
*/
